core/meta/src/TClass.cxx

```diff
@@ -1394,24 +1394,29 @@ void TClass::Init(const char *name, Version_t cversion,
 
    Bool_t isStl = TClassEdit::IsSTLCont(fName);
 
-   if (!gInterpreter) {
+   if (!gInterpreter)
       ::Fatal("TClass::Init", "gInterpreter not initialized");
-   }
 
    if (givenInfo) {
-      if (!gInterpreter->ClassInfo_IsValid(givenInfo) ||
-          !(gInterpreter->ClassInfo_Property(givenInfo) & (kIsClass | kIsStruct | kIsNamespace)) ||
-          (!gInterpreter->ClassInfo_IsLoaded(givenInfo) && (gInterpreter->ClassInfo_Property(givenInfo) & (kIsNamespace))) )
-      {
+      bool invalid = !gInterpreter->ClassInfo_IsValid(givenInfo);
+      bool notloaded = !gInterpreter->ClassInfo_IsLoaded(givenInfo);
+      auto property = gInterpreter->ClassInfo_Property(givenInfo);
+
+      if (invalid || (notloaded && (property & kIsNamespace)) ||
+          !(property & (kIsClass | kIsStruct | kIsNamespace))) {
          if (!TClassEdit::IsSTLCont(fName.Data())) {
             MakeZombie();
             fState = kNoInfo;
             TClass::RemoveClass(this);
             return;
          }
       }
+
+      R__LOCKGUARD(gInterpreterMutex);
```

Comment:
Note that taking the lock inside TClass::Init is (should be) redundant as Init must only be called from the constructors and they all (should) take the lock before calling Init.